src/flb_network.c

```diff
@@ -1167,6 +1187,8 @@ flb_sockfd_t flb_net_tcp_connect(const char *host, unsigned long port,
         freeaddrinfo(res);
     }
 
+    u_conn->busy_flag = FLB_FALSE;
```

Comment:
@leonardo-albertovich Your changes only address part of the bug where the timeout happens before the TCP connection establishment finishes. There are more async calls that happen after the TCP connection establishment - like TLS setup. The busy flag cannot be set to FALSE until the last async call before this line returns: https://github.com/fluent/fluent-bit/blob/master/src/flb_upstream.c#L750 . With current code, it would be https://github.com/fluent/fluent-bit/blob/master/src/flb_io.c#L408 but it might be safer to set the flag to FALSE in flb_upstream.c

Reply:
A simpler change assuming that the coroutine will not yield after `flb_upstream::create_conn` returns is to
- set flag to TRUE before this line: https://github.com/fluent/fluent-bit/blob/master/src/flb_upstream.c#L523
- set it to FALSE here: https://github.com/fluent/fluent-bit/blob/master/src/flb_upstream.c#L537
Otherwise setting/unsetting the flag will be spread all over the code. The above will also address the other scenarios mentioned.
I am still not sure about the TLS cleanup order issue I mentioned in my PR but that could be addressed separately

Reply:
I can modify my PR to include the above if you want to test it out.